cid.go

```diff
@@ -132,36 +132,54 @@ var CodecToStr = map[uint64]string{
 // They exist to allow IPFS to work with Cids while keeping
 // compatibility with the plain-multihash format used used in IPFS.
 // NewCidV1 should be used preferentially.
-func NewCidV0(mhash mh.Multihash) *Cid {
-	return &Cid{
-		version: 0,
-		codec:   DagProtobuf,
-		hash:    mhash,
+func NewCidV0(mhash mh.Multihash) Cid {
+	// Need to make sure hash is valid for CidV0 otherwise we will
+	// incorrectly detect it as CidV1 in the Version() method
+	dec, err := mh.Decode(mhash)
+	if err != nil {
+		panic(err)
+	}
+	if dec.Code != mh.SHA2_256 || dec.Length != 32 {
+		panic("invalid hash for cidv0")
 	}
+	return Cid{string(mhash)}
 }
 
 // NewCidV1 returns a new Cid using the given multicodec-packed
 // content type.
-func NewCidV1(codecType uint64, mhash mh.Multihash) *Cid {
-	return &Cid{
-		version: 1,
-		codec:   codecType,
-		hash:    mhash,
+func NewCidV1(codecType uint64, mhash mh.Multihash) Cid {
+	hashlen := len(mhash)
+	// two 8 bytes (max) numbers plus hash
+	buf := make([]byte, 2*binary.MaxVarintLen64+hashlen)
+	n := binary.PutUvarint(buf, 1)
+	n += binary.PutUvarint(buf[n:], codecType)
+	cn := copy(buf[n:], mhash)
+	if cn != hashlen {
+		panic("copy hash length is inconsistent")
 	}
+
+	return Cid{string(buf[:n+hashlen])}
 }
 
 // Cid represents a self-describing content adressed
 // identifier. It is formed by a Version, a Codec (which indicates
 // a multicodec-packed content type) and a Multihash.
```

Comment:
@kevina Could you please add a high level description of how the `version`, `codec`, and `hash` are packed into a string, similar to how the [multihash docs](https://multiformats.io/multihash/#the-multihash-format) do, to the godoc.

Reply:
We decided against that.  The format varies according to the CID type (see https://github.com/ipfs/go-cid/pull/71#discussion_r213883739).  The layout is according to the cid spec. see https://github.com/ipld/cid.

Reply:
Fair, thoughts on adding a link to the cid spec to the comment then? The location of various specs isn't the easiest thing to deduce, i.e. the cid spec is under the ipld org whereas the implementation is under the ipfs org. Just a suggestion, not a blocker.

Reply:
There entire documentation could use cleaning up. The format is mentioned in multiple places and the link to the spec is in the documentation for the package.  https://godoc.org/github.com/ipfs/go-cid.